Bug 2015493: [OCPCLOUD-1306] User CA bundle sync controller (#136)
openshift/cluster-cloud-controller-manager-operator: docs/dev/trusted_ca_bundle_sync.md

# Managing trusted ca for CCMs

## Intro

> On some, mainly on-prem platforms, such as OpenStack, vSphere, or Azure Stack 
> privately or self-signed SSL certificates might be used for its endpoints.

Reviewer: ```suggestion
On some, mainly on-prem platforms, such as OpenStack, vSphere, or Azure Stack, 
privately or self-signed SSL certificates might be used for its endpoints.
```

> In order to be able to communicate with the underlying platform CA certificates should be
> propagated into CCMs pods trust store.

Reviewer: ```suggestion
In order to be able to communicate with the underlying platform, CA certificates should be
propagated into CCMs pods trust store.
```

Openshift documentation [prescribes](https://docs.openshift.com/container-platform/4.8/networking/configuring-a-custom-pki.html)
to use cluster-wide **_Proxy_** object as a common way to configure trusted CA for an entire cluster.

All necessary logic around `trustedCA` management already exists in Openshift platform, but can not be leveraged by CCM/CCCMO

> in its current state due to CCCMO's place in cluster run sequence.

Reviewer: ```suggestion
in its current state due to CCCMO's role in the cluster bootstrap process.
```

> All necessary steps are performing by [cluster-network-operator](https://github.com/openshift/cluster-network-operator/), which starts on already initialized nodes.

Reviewer: ```suggestion
All necessary steps are performing by [cluster-network-operator](https://github.com/openshift/cluster-network-operator/), which starts only once control plane nodes have been initialized by the CCM.
```

Such node initialization might require communication with the underlying platform (OpenStack, vSphere, ASH),
which would not be successful without configured trust to platform endpoint certificates.

> For solving this 'chicken-and-egg' problem with a minimum amount of risk separate `trusted_ca_bundle_controller` was introduced.

Reviewer: ```suggestion
For solving this 'chicken-and-egg' problem with a minimum amount of risk, a separate `trusted_ca_bundle_controller` was introduced.
```

## Implementation Description

Implementation mostly replicates logic from `cluster-network-operator`.

> The controller has been [introduced](https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/136) as a part of `config-sync-controllers` binary in CCMO pod and lives as separate control loop along with [cloud-config-sync](cloud-config-sync.md) controller. 

Reviewer: ```suggestion
The controller has been [introduced](https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/136) as a part of `config-sync-controllers` binary in CCCMO pod and lives as separate control loop along with [cloud-config-sync](cloud-config-sync.md) controller. 
```
